### DISCUSSION TITLE
Moderator Alerting System

### DISCUSSION THREAD

**Discussion by JamesTCrenshaw** [FakeFishGames/Barotrauma#4740] [Ideas and suggestions]
Currently, there does not exist an alerting system for the moderators/admins of hosted servers. Rather than spending the game glued to the server logs watching for the suspicious actions, there may be configurable alerts for specific actions that would give the admin an alert on-screen. 

Example of desired function: A player begins messing with the wiring on the reactor. The moderator receives an on-screen alert that 'Player X has wired Y to the reactor', thus informing the moderator immediately that a potentially mission-ending grief has occurred. 

Example 2 of desired function: The server host is able to configure the alerting system with a pre-made set of variables. The host can also remove specific players from the alerting system, to avoid false alerts. (Players trustworthy enough to do extremely dangerous events without the intent of griefing.)

Adding this feature, would significantly reduce the amount of events in which servers must restart missions or even entire campaigns due to long-term total ship failure griefs. This is a persistent problem on campaign servers, and most of the time, the griefs have been laid and undetected for such an amount of time that the server had completed a mission, and returned to a station, saving the calamitous and possibly unfixable grief.

**Comment by Rokvach** [comment]
This ties in nicely with an idea I had back in April 2020, I explored the idea more from the UI perspective and how it would appear for the player. 

From our private repo (Link for other devs https://github.com/Regalis11/Barotrauma-development/issues/1687)

> **[Anti-griefing] Popup notifications based on offensive actions**
> Hosts should get a popup notification when a player does something bad. The popup notification should appear above the chat, which stacks up to 3-4 notifications, the oldest notification should fade away after x amount of time has passed.
> 
> Clicking the notification brings up server log at the position the action took place in. The action should be briefly highlighted.
> 
> The system should work without karma being enabled, so if a player attacks another player the host should get a notification of that action, regardless if it was by accident or not.
> 
> The amount of things the host gets notified by should be configurable and not as wide of a range than when having karma enabled.
> 
> So at least the basics should be covered without karma
> 
> Attacked another player
> Unwired a wire
> Plasma cut a wall
> Repeated actions should be stacked to a single notification if no other notifications are received at that time, the notification should indicate that a new notification of the same was received by changing the size of the notification briefly and adding xAmount (x3) at the end of the notification information.

**Reply by NilanthAnimosus** [reply]
I remember when I wrote my own code to handle this entirely server-side only, I ended up offloading it to chat instead of a toast or other system. The issue I always had in my mind for making it client based was where this actually goes on-screen if it were separated from the chat.

For bad actions, I recall checking for things as complicated as inventory insertion into specific item types (IE. welding fuel inside diving suit), C4 into detonators. rather than just the pickup of a detonator (Unless it contained a blacklisted item already) or cutting of a wall.

I made mine horribly complex, and the game definitely was not designed for it (Picture networking code for items and me trying to monitor each different place that triggers for 5 different kinds of event). Some areas not even including the character/client responsible and having to forward the information on.

I like the sound of the popup notification and stacking it, but if say, the host is also the captain, whom has bots. whats a good location on the screen that is not too obtrusive to game play to list such actions in a sort of toast stack? many hosts play and game, as well as admins really.
